### ISSUE TITLE
Evaluate the possibility of using an event handler based system, not a republisher based system

### ISSUE THREAD

**Issue by schomatis** [ipfs/go-mfs#33]
See https://github.com/ipfs/go-ipfs/issues/5738#issuecomment-438048895.

**Comment by schomatis** [comment]
@Stebalien Could you expand some more on how that would look like in the code?

**Comment by Stebalien** [comment]
Basically, I'd like a way to *subscribe* to changes to a file/directory and then act on them. A debounced republisher would be *an* event handler (something that subscribes to changes) but not the only kind.

This would also get rid of the concept of a "root" (and potentially get rid of "parents").

However, we can probably fix the issue by starting with something less invasive. My real goal here is to get rid of root/republisher.

**Comment by schomatis** [comment]
> A debounced republisher would be _an_ event handler

What we got now is a debounced republisher, right? Then we would spawn a republisher goroutine in any `FSNode` the user subscribes to?

> we can probably fix the issue by starting with something less invasive.

Do you have any suggestions about less invasive approaches? I'm a bit lost here regarding what we're aiming for.

**Comment by Stebalien** [comment]
**Events**

This was mostly me just trying to figure out a nice way to fix this issue, get rid of "childClosers", republishers, the root, etc. all in one go.

1. Make a generalized event system.
2. Provide a small adapter for debouncing events from an FSNode. This would allow us to replace the republisher.

However, let's just punt on that.

**The Bug**

The real bug, as far as I can tell, is that we're flushing asynchronously. While it may make sense to coalesce a few events, `Flush()` shouldn't return until the new root CID has been written to disk. Any solution that fixes that bug works.

**Comment by schomatis** [comment]
> `Flush()` shouldn't return until the new root CID has been written to disk.

I think the orthodox solution here would be then to call `WaitPub` after `Update` in the root (and fixing `WaitPub` - https://github.com/ipfs/go-mfs/issues/38).

----------------------------------------



> * Make a generalized event system.

Something like https://github.com/olebedev/emitter? (not that we need to use it, but I mean, that kind of architecture?)

**Comment by Stebalien** [comment]
> I think the orthodox solution here would be then to call WaitPub after Update in the root (and fixing WaitPub - #38).

Currently, that's the best we can do. But, as a programmer, I'd expect `Flush` to flush.

> Something like https://github.com/olebedev/emitter? (not that we need to use it, but I mean, that kind of architecture?)

Yes but we can likely get away with something a bit simpler.

**Comment by schomatis** [comment]
Blocked on #38 (let's fix the `WaitPub` logic before taking care of this).

**Comment by aschmahmann** [comment]
This repository is no longer maintained and has been copied over to [Boxo](https://github.com/ipfs/boxo).  In an effort to avoid noise and crippling in the Boxo repo from the weight of issues of the past, we are closing most issues and PRs in this repo.  Please feel free to open a new issue in Boxo (and reference this issue) if resolving this issue is still critical for unblocking or improving your usecase.

You can learn more in the [FAQs for the Boxo repo copying/consolidation effort](https://github.com/ipfs/boxo/wiki/Copied-or-Migrated-Repos-FAQ).
